minio/minio
# Keep an up-to-date copy of the KMS master key

- Keep an up-to-date copy of the KMS master key
- Validate KMS master key using a test of encrypting decryption, in order to startup MinIO

## Community Contribution License
All community contributions in this pull request are licensed to the project maintainers
under the terms of the [Apache 2 license](https://www.apache.org/licenses/LICENSE-2.0). 
By creating this pull request I represent that I have the right to license the 
contributions to the project maintainers under the Apache 2 license.

## Description
MinIO should automatically request the default master key once every 10 seconds. This matches the KES period for verifying if the KMS is offline. The intention is to have the master key cached, so that the IAM and other cluster metadata can be decrypted even if the KMs goes offline

Currently when the KMS goes offline, if MinIO is restarted for some reason it immediately tests its ability to create a key through KES on the KMS. Since KES cannot contact the KMS, MinIO is unable to start up and responds with:
`ERROR Unable to initialize a connection to KES as specified by the shell environment: failed to create key`

When the KMS is down, KES does not garbage collect any cache, unless `.cache.expiry.offline` is setup. Therefore in order for MinIO to start up, a way was needed to validate that the cached KMS key can be used to encrypt and decrypt.


## Motivation and Context
This is the precise scenario:
* MinIO starts and decrypts its IAM data using the key MINIO_KMS_KES_KEY_NAME. Therefore, KES has to fetch the key from the KMS backend (if not cached already). 

* After some time KES evicts the key MINIO_KMS_KES_KEY_NAME from the cache since its not used by MinIO. See usage of `.cache.expiry.unused`. 
`.cache.expiry.any` is out of scope since it always purges all cache if KES verifies the KMS to be online. This is a race condition.
`.cache.expiry.offline` is also out of scope since it always purges all cache if KES verifies the KMS is offline. This is a race condition.

* The KMS becomes temp. unavailable (network issue, maintance etc.) 

* MinIO gets restarted. This now fails because KES cannot find the MINIO_KMS_KES_KEY_NAME in its cache and it also cannot fetch it from the KMS backend because the KMS is not available.

This fix allows MinIO to start up successfully, with most of its functionality enabled except creating a new KMS key, nor encryption a bucket with cached KMS keys. Only data encrypted with the default key is accessible.

Currently KES listKeys does not use its cache when it fails to access the KMS (https://github.com/minio/kes/blob/master/keystore.go#L342). Therefore, this fix is limited in its scope and while it does permit disabling encryption on future upload, it does not allow a user to enable encryption on buckets while the KMS is down.

## How to test this PR?
See https://github.com/allanrogerr/public/wiki/mineos%E2%80%907
Note: Increase `any` and `offline` to large values to reduce test scope

### Summary
#### Test 0
```
cache:
  expiry:
    any: 600s
    unused: 30s
    offline: 600s
```
* Start vault, verify KES running, start MinIO, create and encrypt bucket with default key, upload data, kill vault, restart minio after 30s
* Observe minio starts up
* Validate existing minio encrypted data is accessible
* Validate new minio encrypted data is uploadable and accessible
* Validate encryption can be removed and new minio data is uploadable and accessible


#### Test 1
```
cache:
  expiry:
    any: 600s
    unused: 20s
    offline: 600s
```
* Start vault, verify KES running, start MinIO, create and encrypt bucket, upload data, kill vault, restart minio after 10s
* Observe minio starts up sometimes due to the inherent race condition
* Validate existing minio encrypted data is accessible
* Validate new minio encrypted data is uploadable and accessible
* Validate encryption can be removed and new minio data is uploadable and accessible


#### Test 2
```
cache:
  expiry:
    any: 600s
    unused: 10s
    offline: 600s
```
* Start vault, verify KES running, start MinIO, create and encrypt bucket, upload data, kill vault, restart minio after 10s
* Observe minio starts up sometimes due to the inherent race condition


#### Test 3
```
cache:
  expiry:
    any: 600s
    unused: 5s
    offline: 600s
```
* Start vault, verify KES running, start MinIO, create and encrypt bucket, upload data, kill vault, restart minio after 5s
* Observe minio does not start up because KES cache was purged before it could be refreshed by minio

## Types of changes

## Checklist:
